### PR TITLE
Feature/webos player

### PR DESCRIPTION
**Pull request by arokor**
Add webos player MVP.
All type definitions are AI generated from the documentation

### CODE REVIEW

File: packages/device/src/components/basemenu.component.tsx [hunk 1]
```diff
@@ -197,21 +197,28 @@ export const BaseMenu: Component<BaseMenuProps> = ({
...
+
   // register event listeners. We may need more generic key events in the future to
   // support other devices. TV remote for example.
   onMount(() => {
     const onKeyDown = (e: KeyboardEvent) => {
       showMenu();
 
       const currentItems = items();
-      if (e.key === 'ArrowUp') {
+      if (e.keyCode === Key.Up) {
```
Comment (manast) [inline packages/device/src/components/basemenu.component.tsx:214]: Why not using a switch statement here?

File: packages/platforms/webos-player/src/classes/file-storage.test.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    });
+  });
+
+  describe('Downloading', () => {
+    it('should download a file from a URL and write it to the file system', async () => {
+      vi.mocked(api.statFile).mockResolvedValueOnce({ size: 0 });
+      const url = 'http://example.com/image.png';
+      const result = await storage.storeFile(url);
+
+      console.log('aa', result);
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/file-storage.test.ts:103]: debug console

File: packages/platforms/webos-player/src/classes/file-storage.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+export class FileStorage implements StorageIntegration {
+  private storagePath: string;
+
+  constructor(storagePath: string) {
+    this.storagePath = storagePath;
+  }
+
+  async init(): Promise<void> {
+    console.log('Initializing file storage');
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/file-storage.ts:26]: debug console

File: packages/platforms/webos-player/src/classes/file-storage.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    return storage.mkdir({ path: CACHE_PATH });
+  }
+
+  async info(): Promise<StorageInfo> {
+    console.log('Getting storage info');
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/file-storage.ts:31]: debug console

File: packages/platforms/webos-player/src/classes/file-storage.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    try {
+      const { files } = await storage.listFiles({ path: CACHE_PATH });
+      let used = files.reduce((acc, file) => acc + (file.size ?? 0), 0);
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/file-storage.ts:34]: let should be const

File: packages/platforms/webos-player/src/classes/webos-machine.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  appType: 'ipk',
+  fqdnMode: true,
+  fqdnAddr: 'https://update.castmill.io/webos/player-new.ipk',
+};
+
+var CREDENTIALS_FILE_PATH = 'credentials.txt';
+
+export class WebosMachine implements Machine {
+  async getMachineGUID(): Promise<string> {
+    console.log('getMachineGUID');
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/webos-machine.ts:22]: debug console

File: packages/platforms/webos-player/src/classes/webos-machine.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+      throw new Error('No mac address found');
+    }
+
+    const hash = await digestText(macAddress);
+
+    return hash;
+  }
+
+  async storeCredentials(credentials: string): Promise<void> {
+    console.log('storeCredentials', credentials);
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/webos-machine.ts:36]: debug console

File: packages/platforms/webos-player/src/classes/webos-machine.ts [hunk 8]
```diff
@@ -0,0 +1,184 @@
+import { Machine, DeviceInfo } from '@castmill/device';
+import { configuration, deviceInfo, power, storage } from '../native';
+import { digestText } from './utils';
+import { version } from '../../package.json';
+
+// The update config is used to update the application on the device. We set this
+// to make sure the application updates from the correct source.
+var UPDATE_CONFIG = {
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/webos-machine.ts:8]: should use const

File: packages/platforms/webos-player/src/classes/webos-machine.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+  serverIp: '0.0.0.0',
+  serverPort: 0,
+  secureConnection: true,
+  appLaunchMode: 'local',
+  appType: 'ipk',
+  fqdnMode: true,
+  fqdnAddr: 'https://update.castmill.io/webos/player-new.ipk',
+};
+
+var CREDENTIALS_FILE_PATH = 'credentials.txt';
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/webos-machine.ts:18]: should use const

File: packages/platforms/webos-player/src/classes/webos-machine.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+  /**
+   * Updates the device's firmware.
+   */
+  async updateFirmware(): Promise<void> {
+    console.log('updateFirmware');
+
+    const url = await this.getFirmwareDownloadUrl();
+
+    await storage.downloadFirmware({
```
Comment (manast) [inline packages/platforms/webos-player/src/classes/webos-machine.ts:168]: Will this file be removed automatically after the upgrade?

**Review by manast — APPROVED** [review]
LGTM, there are a few console.logs though in several places.

File: packages/device/src/components/basemenu.component.tsx (hunk 1 above)
Comment (arokor) [reply packages/device/src/components/basemenu.component.tsx]: Hmm. No idea, fixed.

File: packages/platforms/webos-player/src/classes/file-storage.test.ts (hunk 2 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/file-storage.test.ts]: fixed

File: packages/platforms/webos-player/src/classes/file-storage.ts (hunk 4 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/file-storage.ts]: fixed

File: packages/platforms/webos-player/src/classes/file-storage.ts (hunk 5 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/file-storage.ts]: fixed

File: packages/platforms/webos-player/src/classes/webos-machine.ts (hunk 6 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/webos-machine.ts]: fixed

File: packages/platforms/webos-player/src/classes/webos-machine.ts (hunk 8 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/webos-machine.ts]: fixed

File: packages/platforms/webos-player/src/classes/webos-machine.ts (hunk 9 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/webos-machine.ts]: fixed

File: packages/platforms/webos-player/src/classes/webos-machine.ts (hunk 10 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/webos-machine.ts:154]: Yes. It is deleated automatically upon reboot. Added a comment about it

File: packages/platforms/webos-player/src/classes/webos-machine.ts (hunk 7 above)
Comment (arokor) [reply packages/platforms/webos-player/src/classes/webos-machine.ts]: fixed

**Pull request merged**